include/oneapi/dpl/pstl/execution_defs.h

```diff
@@ -201,13 +201,60 @@ template <typename _ExecPolicy, typename _T>
 using __ref_or_copy =
     typename oneapi::dpl::__internal::__ref_or_copy_impl<typename ::std::decay<_ExecPolicy>::type, _T>::type;
 
+// utilities for Range API
 template <typename _R>
-using __difference_t = decltype(::std::declval<_R&>().size());
+auto
+__check_size(int) -> decltype(::std::declval<_R&>().size(), ::std::true_type{});
```

Comment:
I believe I have the idea, how to simplify all of this stuff. Let's discuss offline

Reply:
The approach we agreed on is proven there: https://godbolt.org/z/c34s96